### ISSUE TITLE
Request to change currentTarget in Event interface for lib.d.ts

### ISSUE THREAD

**Issue by nathggns** [microsoft/TypeScript#299]
When using the Event interface from lib.d.ts, and attaching a listener, the callback will get an object of type Event. However, the Event's currentTarget property is of type EventTarget (whereas it's should be of type Element/HTMLElement).

**Comment by rayshan** [comment]
Hi all, pinging this thread. I ran into this issue with `Event.target`. I would imagine most of the time a target is an `HTMLElement`. It is a little odd that the default type assumes it isn't. I tried @SaschaNaz's suggestion and ran into the "duplicate identifier" issue (sorry I may be missing something, I'm pretty new to TypeScript).

**Comment by kitsonk** [comment]
@rayshan the suggestion was for what should go into the `lib.d.ts` file and the file is auto generated based directly on the implementation of Internet Explorer.  So the error you encountered is accurate.  Right now the only way around it is to build without the built in lib (`--noLib`) and use the workaround, or do what you have done, which is cast the target.  There are currently a number of challenges of dealing with the return types of the DOM right now in TypeScript (mostly because the DOM isn't the most consistent/structured/type safe/consistently implemented set of APIs).

**Comment by SaschaNaz** [comment]
Now that #4910 is merged, can this be revisited?

**Comment by danquirk** [comment]
@zhengbli might have some context on the status of lib.d.ts issues that require script based generation

**Comment by mhegazy** [comment]
PRs welcomed. here is some infromation on contributing lib.d.ts changes: https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md#contributing-libdts-fixes

**Comment by devdanke** [comment]
I ran into this `TS2339: Property 'result' does not exist on type 'EventTarget'` in JS FileReader `onload`, and another warning for `getSummary()` on the event passed to FileReader's `onerror`.

My work-around, to suppress the horrid red squiggily lines;-) is the following:

```
interface FileReaderEventTarget extends EventTarget {
    result:string
}

interface FileReaderEvent extends Event {
    target: FileReaderEventTarget;
    getMessage():string;
}
```

Then in my app:
```
reader.onload = function(fre:FileReaderEvent) {
        var data = JSON.parse(fre.target.result);
        ...
    }
```

The generated JS codes looks good and works for me.  Could this kind of solution, for unusual JS event types, be added to lib.d.ts?  

I'm guessing my solution is overly simplistic.  But it would help to know why.

**Comment by DaSchTour** [comment]
I would love to fix this issue. But as I see the only reference to currentTarget is inside generated webworker.generated.d.ts files in TSJS generator. I have no idea how to fix this 😞 
I really hope somebody could fix this nearly 3 years old issue. Such errors may be the reason why I often hear people say that TypeScript is hard to use and that really makes me sad.

**Comment by DaSchTour** [comment]
Well. I tried to start solve this issue. I create a PR https://github.com/Microsoft/TSJS-lib-generator/pull/202
Hopefully this is going into the right direction.

**Comment by ccorcos** [comment]
I'm getting `Property 'getBoundingClientRect' does not exist on type 'EventTarget'.`. I think this is a good idea.

**Comment by DaSchTour** [comment]
Well, the PR is pending since nearly 2 month. I'll look into this if there is anything that could be done to speed this up.

**Comment by felixfbecker** [comment]
It would be really great if we could have `currentTarget` be generic, so that for example `HTMLInputElement.onchange` can be of type `(event: Event<HTMLInputElement>) => void` so that `event.currentTarget.value` works without any cast or type annotations.

**Comment by SaschaNaz** [comment]
This is blocked by performance issue caused by generics as introducing generics doubles memory usage 😭

**Comment by felixfbecker** [comment]
Do these performance problems apply equally to the React typings?

**Comment by SaschaNaz** [comment]
lib.d.ts doesn't include React typings but you can test it by adding `--diagnostics` to the command.

**Comment by DaSchTour** [comment]
@felixfbecker I personally doubt that this will be fixed soon. I guess most people already created their workaround or abandoned TypeScript. That's also why there are so few people hitting this problem.

**Comment by ywang04** [comment]
This issue still exists in Version 1.18.0 (1.18.0)

**Comment by anouarcharif** [comment]
Hello,

Same issue with ... and i guess it's not fixed yet.

**Comment by zheeeng** [comment]
Hope some progress on this issue. It is hard to manipulate dom with Typescript.

**Comment by DaSchTour** [comment]
This issue is four years old and the corresponding PR nearly one year I doubt this will be fixed any time soon.

**Comment by MrKou47** [comment]
This issue has been around for 4 years but has not been fixed yet.

**Comment by styfle** [comment]
There is a fix available in https://github.com/Microsoft/TSJS-lib-generator/pull/207
But @saschanaz is blocking based on the performance.

**Comment by guntram** [comment]
Casting party tonite at 11 - unload your browser tabs:
`confirmationMessage(event: BeforeUnloadEvent): any {
    const activeElement: HTMLElement = <HTMLElement>(<Document>event.target).activeElement;`

:dancing_men:

**Comment by HolgerJeromin** [comment]
What about adding 

```ts
    readonly currentTarget: Element | null;
    readonly target: Element | null;
```
on `UIEvent` only? This is not very specific, but makes most code happy.

**Comment by mikewodarczyk** [comment]
It seem like I can write this kind of code (see below) which seems to behave as if HTMLElement and its friends are classes.  I have to call them as "new MyDiv();" instead of "new HTMLDivElement()", but it seems as if the library that gives me "HTMLDivElement" could have done this for me.   Why shouldn't Typescript change the semantics of "document.createElement("foo") and HTMLFooElement" into something more Typescript-like.  The language is all about changing javascript semantics.  

    export class MyDiv extends HTMLDivElement {
       constructor() {
        try {
            super();
        }
        finally {
            return document.createElement("div") as HTMLDivElement;
        }
      }
    }

    export class MyInp extends HTMLInputElement {
      constructor(value: string) {
        try {
            super();
        }
        finally {
            const x = document.createElement("input") as HTMLInputElement;
            x.value = value;
            return x;
        }
      }
    }


    export class TestElement extends HTMLElement {
       constructor() {
        try {
            super();
        }
        finally {
            let x = new MyDiv();
            let inp = new MyInp("xyzzy");
            x.appendChild(inp);
            return x;
        }
      } 
    }

**Comment by simonh1000** [comment]
same issue here. I was trying to use Vue with Typescript and wanted to get the value of an input field as the user typed. But event.target.value would not pass the type checker even though - i think - a text inout must always produce such fields for such an event

**Comment by gautamkrishnar** [comment]
We got rid of the error by adding type any to our code:
```ts
solution : this.url = (<any>event).target.result;
```

**Comment by pavethiran** [comment]
What about typing Event and EventTarget?

```ts
interface Event<C = any, S = any, T = any> {
  ...
  currentTarget: EventTarget<C> | null;
  srcElement: EventTarget<S> | null;
  target: EventTarget<T> | null;
  ...
}

interface EventTarget<T = any> {
  ...
}
```

**Comment by adigourdi** [comment]
I encountered this issue while using the FileReader, a simple cast to the correct type fixes it

```typescript
const fileReader = new FileReader();
fileReader.onload = $ev => {
  console.log($ev); // type ProgressEvent
  console.log($ev.target); // type FileReader
  // console.log($ev.target.result); // editor and autocomplete doesn't show any error
  console.log(($ev.target as FileReader).result); // casting compiles fine
};
fileReader.readAsText(file);
```

**Comment by alexandercerutti** [comment]
Are there any updates about this topic? It would be useful to have it without force casting `event.target` every time. Also this is open since 2014. Thank you!

**Comment by jonmol** [comment]
Indeed, would be nice with some kind of official response to this issue before its 6th birthday

**Comment by mindplay-dk** [comment]
Yep, this is probably the Number 1 day-long nuisance when working with DOM, JSX, etc.

In fact, it's the only daily nuisance I have - and I imagine a million people have this, all day long.

It's really difficult to understand why this doesn't get any priority.

**Comment by Dean-NC** [comment]
Same for event.target.  A very common scenario is adding a click handler to a parent element when children don't exist yet, and when there will be many children. A simple test for event.target.tagName or even classList is a common thing I do. It works....have used it for many years.  I don't get intellisense for tagName/classList since EventTarget is not classified as HTMLElement, but it works.

**Comment by allejo** [comment]
> This is blocked by performance issue caused by generics as introducing generics doubles memory usage 😭

Going to bump this thread since the last post was in 2020. Are the performance issues with generics still there? Is there a concern about introducing performance issues to users with old versions of TS if modern TS can handle things well? Is there anything the community can do? Would a PR be welcome?

**Comment by minecrawler** [comment]
What I can see is that there is interest in fixing this issue, and there was a PR. The only reason the PR was not merged is because its performance was bad (because of generics). Then the PR was closed, because it wasn't updated anymore.

Now, people are asking for this feature, but there is no new PR with new performance numbers. I don't feel comfortable to implement this feature since I never worked on tsc, though is there anyone else who is able to do so? From the old PR, it doesn't look like it would be a lot of work.

**Comment by nscarcella** [comment]
For anyone still interested in this feature, I created a PR [here](https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/1624) and could use some feedback regarding the performance.
